rmw_implementation/src/functions.cpp

```diff
@@ -0,0 +1,513 @@
+#include <string>
+
+#include <fstream>
+#include <sstream>
+
+#include "Poco/SharedLibrary.h"
+
+#include "rmw/rmw.h"
+
+std::string get_env_var(const char * env_var)
+{
+  char * value = nullptr;
+#ifndef _WIN32
+  value = getenv(env_var);
+#else
+  size_t value_size;
+  _dupenv_s(&value, &value_size, env_var);
+#endif
+  std::string value_str = "";
+  if (value) {
+    value_str = value;
+#ifdef _WIN32
+    free(value);
+#endif
+  }
+  // printf("get_env_var(%s) = %s\n", env_var, value_str.c_str());
+  return value_str;
+}
+
+std::list<std::string> split(const std::string & value, const char delimiter)
+{
+  std::list<std::string> list;
+  std::istringstream ss(value);
+  std::string s;
+  while (std::getline(ss, s, delimiter)) {
+    list.push_back(s);
+  }
+  // printf("split(%s) = %zu\n", value.c_str(), list.size());
+  return list;
+}
+
+bool is_file_exist(const char * filename)
+{
+  std::ifstream h(filename);
+  // printf("is_file_exist(%s) = %s\n", filename, h.good() ? "true" : "false");
+  return h.good();
+}
+
+std::string find_library_path(const std::string & library_name)
+{
+  const char * env_var;
+  char separator;
+  const char * filename_prefix;
+  const char * filename_extension;
+#ifdef _WIN32
+  env_var = "PATH";
+  separator = ';';
+  filename_prefix = "";
+  filename_extension = ".dll";
+#elif __APPLE__
+  env_var = "DYLD_LIBRARY_PATH";
+  separator = ':';
+  filename_prefix = "lib";
+  filename_extension = ".dylib";
+#else
+  env_var = "LD_LIBRARY_PATH";
+  separator = ':';
+  filename_prefix = "lib";
+  filename_extension = ".so";
+#endif
+  std::string search_path = get_env_var(env_var);
+  std::list<std::string> search_paths = split(search_path, separator);
+
+  std::string filename = filename_prefix;
+  filename += library_name + filename_extension;
+
+  for (auto it : search_paths) {
+    std::string path = it + "/" + filename;
+    if (is_file_exist(path.c_str())) {
+      return path;
+    }
+  }
+  return "";
+}
+
+#define STRINGIFY_(s) #s
+#define STRINGIFY(s) STRINGIFY_(s)
+
+Poco::SharedLibrary *
+get_library()
+{
+  static Poco::SharedLibrary * lib = nullptr;
+  if (!lib) {
+    std::string env_var = get_env_var("RMW_IMPLEMENTATION");
+    if (env_var.empty()) {
+      env_var = STRINGIFY(DEFAULT_RMW_IMPLEMENTATION);
+    }
+    std::string library_path = find_library_path(env_var);
+    if (library_path.empty()) {
+      return nullptr;
+    }
+    lib = new Poco::SharedLibrary(library_path);
+  }
+  return lib;
+}
+
+void *
+get_symbol(const char * symbol_name)
+{
+  Poco::SharedLibrary * lib = get_library();
+  if (!lib->hasSymbol(symbol_name)) {
+    return nullptr;
+  }
+  return lib->getSymbol(symbol_name);
+}
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#define ARG_TYPES(...) __VA_ARGS__
+#define ARG_VALUES(...) __VA_ARGS__
+
+#define CALL_SYMBOL(symbol_name, ReturnType, error_value, ArgTypes, arg_values) \
+  void * symbol = get_symbol(symbol_name); \
+  if (!symbol) { \
+    return error_value; \
```

Comment:
Good point. I added rmw error messages for two cases and also a try / catch block around the library loading: see 12e12f17e3037c3c60a1f5ca03739b8061eeae6b